wagtail/core/migrations/0063_collection_management_permissions.py

```diff
@@ -3,60 +3,54 @@
+    groups_w_permissions = Group.objects.filter(
+        permissions__content_type__app_label='wagtailcore',
+        permissions__content_type__model='collection',
+        permissions__codename__in=['add_collection', 'change_collection', 'delete_collection']
+    ).values('id', 'name', 'permissions__id', 'permissions__codename')
 
-    for perm in stock_permissions:
-        for group in perm.group_set.all():
-            gcp = GroupCollectionPermission(
-                group_id=group.id,
+    for root_collection in Collection.objects.filter(depth=1).all():
+        for row in groups_w_permissions:
+            GroupCollectionPermission.objects.create(
+                group_id=row['id'],
+                permission_id=row['permissions__id'],
                 collection_id=root_collection.id,
-                permission_id=perm.id
             )
-            gcp.full_clean()
-            gcp.save()
-            # Now remove the model-level permissions from the group
-            group.permissions.remove(perm)
+    # Now remove the model-level permissions for collections
+    collection_permissions = Permission.objects.filter(
+        content_type__app_label='wagtailcore',
+        content_type__model='collection',
+        codename__in=['add_collection', 'change_collection', 'delete_collection'],
+    )
+    for perm in collection_permissions.all():
+        perm.group_set.clear()
 
 
 def revert_to_model_level_collection_management_permissions(apps, schema_editor):
+    """
+    Give model-level permission to all groups who have that permission on the root collection
+    """
     Collection = apps.get_model('wagtailcore.Collection')
-    ContentType = apps.get_model('contenttypes.ContentType')
     GroupCollectionPermission = apps.get_model('wagtailcore.GroupCollectionPermission')
-    Permission = apps.get_model('auth.Permission')
 
-    collection_content_type = ContentType.objects.get(
-        model='collection',
-        app_label='wagtailcore'
-    )
-    root_collection = Collection.objects.filter(depth=1).order_by('path').first()
-    stock_permissions = Permission.objects.filter(
-        content_type__app_label='wagtailcore',
-        content_type__model='collection',
-        codename__in=['add_collection', 'change_collection', 'delete_collection']
-    ).all()
-
-    # Give model-level permissions to all groups who have permission on the root collection
+    root_collections = Collection.objects.filter(depth=1).all()
     group_collection_permissions = GroupCollectionPermission.objects.filter(
-        collection=root_collection,
-        permission__in=stock_permissions
-    ).prefetch_related('group')
+        permission__content_type__app_label='wagtailcore',
+        permission__content_type__model='collection',
+        permission__codename__in=['add_collection', 'change_collection', 'delete_collection'],
+        collection__in=root_collections,
+    ).prefetch_related('group', 'permission')
```

Comment:
Since the `Group` and `Permission` objects are pretty tiny (in memory usage terms), I think we could safely get away with `select_related()` here instead. Doing so should reduce this from 3 database queries to 1.

Reply:
Ahh good catch. Corrected.